### ISSUE TITLE
24 hour telemetry window & user privacy

### ISSUE THREAD

**Issue by morganherlocker** [openmobilityfoundation/mobility-data-specification#282]
A recent change on the `0.3.x` branch added a requirement to the Agency spec that asks location telemetry to be POST'd to the agency within 24 hours. My understanding from communicating with contributors to this spec offline is that **user location telemetry with sub-24h delay was deemed unnecessary to achieve enforcement and policy goals required by cities, so the change was made to provide a level of privacy protection for citizens, in line with [City of Los Angeles' new data protection requirements](https://ladot.io/wp-content/uploads/2019/03/LADOT_Data_Protection_Principles-1.pdf) for micromobility**. I 100% support the new data protection policy, and I agree with the intent of protecting user privacy with a 24 hour delay to protect safety, but I noticed a few gaps that still leak real time location data of a sensitive nature.

https://github.com/CityOfLosAngeles/mobility-data-specification/commit/18978b384b91855718aa055b687480832eb32c81#diff-1d1d5778763061ebb2bdc5db696077a6

- No changes were made to the Provider specification, which still contain multiple forms of real time location data. An agency is free to scrape the provider in real time and the provider is still required to return real time data without delay.
- The change was only made to the telemetry endpoint, which handles certain types, but not all location data. The most vulnerable data to re-identification is origin destination trip data, which has not been modified from its real time requirement. 
- The "push **OR** pull" structure to the spec adds significant ambiguity around these types of issues. A change on Agency spec must be mirrored to the Provider spec, otherwise the same sensitive data is exposed. If MDS is intentionally shifting to a 24 hour delay to protect privacy, this needs to be described across **all endpoints revealing real time user location**.
- The real time origin and destinations present a significant citizen safety risk. For example, an attacker could monitor the feed for a moment when a trip starts (via GBFS or status_changes feed) at a sensitive location, then wait for a corresponding trip end event. If the end of the trip is residential, the identity of the person would be revealed through public sources (phonebook). For example, this would expose the identities of anyone visiting a sensitive medical facility or place of worship, and would allow a malicious party to know when they are home. A 24 hour delay would still allow for origin destination re-identification, but would at least prevent the victim's **current location** from being revealed.

**Comment by xxmarcelxx** [comment]
Morgan,

Thanks for your questions. A few clarifying questions to help us better understand:

1. Is your primary concern with public access to data related to trip origin / destination (O/D) data? Or, is it with City DOTs having this information? Or both?

2. On the public side, it appears your concern relates primarily to GBFS since MDS is not a public feed. If yes, this echoes concerns raised by JUMP in issue #75. That issue proposed that vehicle ID should be randomized between trips to make it harder to determine O/D pairs from cached GBFS data. We have been led to believe that many mobility providers already do this in practice. Would this address your public access concerns?

3. In terms of a City accessing the MDS Provider API, there is not currently a latency requirement (time between event occurring and event appearing in the feed) specified by the MDS Provider definition. In practice, feeds are definitely not real-time, with a latency often measured in hours. Although the spec is silent, a city could conceivably write a latency requirement into policy. Are you proposing a minimum latency requirement in the spec itself? (e.g. MDS Provider records returned to a city that show a vehicle location history must be at least 24 hours old). 

Thank you.

Marcel

**Comment by morganherlocker** [comment]
Thanks for the thoughtful followup @xxmarcelxx 🙇 

## QAs

> 1. Is your primary concern with public access to data related to trip origin / destination (O/D) data? Or, is it with City DOTs having this information? Or both?

**Cities need the information communicated by MDS to effectively regulate the providers in their jurisdiction.** This issue is written with the assumption that the agency 1. has a legal right to the data and a policy necessity and 2. is operating MDS infrastructure with appropriate security and [access protocols](https://en.wikipedia.org/wiki/Principle_of_least_privilege) in place. The details of 2 are important, but out of scope for this issue, which assumes we are all in agreement on what it means to handle MDS data safely.

> 2. On the public side, it appears your concern relates primarily to GBFS since MDS is not a public feed. If yes, this echoes concerns raised by JUMP in issue #75. That issue proposed that vehicle ID should be randomized between trips to make it harder to determine O/D pairs from cached GBFS data. We have been led to believe that many mobility providers already do this in practice. Would this address your public access concerns?

Agree, **raw** MDS data should not be made public, as the location/time portions of the feed are PII under any reasonable definition. Some cities are publishing MDS data to the public via side channels, with modifications to the data and format intended to obscure individual behavior. I believe everyone is in agreement that this is separate from MDS protocol itself, and should only be done in ways that offer meaningful and consistent privacy protection. Considering the rising adoption of MDS, I do think a detailed recommendation for how to open transformed MDS data safely would be a responsible, if optional, thing to do. For this issue, I am assuming internal, appropriately scoped access as noted above.

The 24 hour delay has been widely reported on as a privacy protection, referencing [presentation documents separate from this spec](https://ladot.lacity.org/sites/g/files/wph266/f/MDS%20Developer%20Webinar%20-%20One%20Year%20Permitting%20Overview_03-06-19_REVISION.pdf) that include technical details that have not yet been implemented in MDS itself. I'm not weighing in on whether the delay is necessary or sufficient to protect privacy, however, the intent impacts how engineers will build around MDS and how riders will understand the privacy protections in place. My primary concern here is the technical implications of the intent, and the privacy implications are incidental. As an engineer building on MDS, it is critical that communications about the spec do not diverge in any way from the spec itself, creating fragmentation. This is an engineering necessity to make things work reliably for agencies and providers.

> 3. In terms of a City accessing the MDS Provider API, there is not currently a latency requirement (time between event occurring and event appearing in the feed) specified by the MDS Provider definition. In practice, feeds are definitely not real-time, with a latency often measured in hours. Although the spec is silent, a city could conceivably write a latency requirement into policy. Are you proposing a minimum latency requirement in the spec itself? (e.g. MDS Provider records returned to a city that show a vehicle location history must be at least 24 hours old).

This describes my technical & privacy concerns well. **If the intent is to provide 24 hour delay as meaningful privacy guarantee to users about their real time location, the spec should explicitly disallow it.** If that is the intent, there is also no need to have the Agency POST endpoint for pushing telemetry, since it is redundant to the Provider historical GET endpoint. If there is a true across-the-board 24 hour delay, a city could run a quick batch import of the previous day's data, which would be more secure and far cheaper/easier to maintain. Something like the min latency requirement you mention would cover this.

**If the 24 hour delay is not intended as a privacy protection, and simply a max latency guarantee for engineering purposes, then providers and agencies could still share data in real time, and a public correction should be made clarifying the intent.** Agencies with authority do have the legal right to access real time location telemetry, however, agencies should communicate explicitly whether or not they are choosing to exercise this right. If this is the case, the existing API with potential for real time data ingestion is designed correctly, and the only changes I would propose are related to communications including documentation.

**Comment by HenriJ** [comment]
As the "24 hours" have been removed from the specification, should we close this issue ?
Or we might keep it as an open-discussion whether it is the responsibility of the specification to define the accepted latency or whether it should be defined directly by cities on a case-by-case basis ?

@hunterowens @morganherlocker ?

**Comment by morganherlocker** [comment]
> As the "24 hours" have been removed from the specification, should we close this issue ?

The context at the time of this post was that there were numerous press statements touting the 24 hour delay as a privacy and user safety feature. I opened this issue when I noticed gaps in the spec that would prevent this from being an effective privacy mitigation strategy. It is now clear that these statements were not an accurate representation of the content or intent of the specification. While I still believe real time user surveillance is a serious safety issue, the questions that prompted this disclosure have been addressed, so I will close.
